cassiopeia/solver/graph_utilities.py

```diff
@@ -160,3 +160,213 @@ def max_cut_improve_cut(G: nx.DiGraph, cut: List[int]):
+                for v in G.neighbors(u)
+                if check_if_cut(u, v, new_cut)
+            ]
+        )
+        delta_num[u] = d - 2 * c
+        if u in new_cut:
+            delta_denom[u] = -d
+        else:
+            delta_denom[u] = d
+        # Set the improvement potential for each node
+        set_improvement_potential(u)
+
+    all_pos = False
+    iters = 0
+
+    while (not all_pos) and (iters < len(G.nodes)):
+        # Negative potentials decrease the cut, so iterates until moves would
+        # no longer decrease the cut, thus minimizing
+        best_potential = min(improvement_potentials.values())
+        if best_potential < 0:
+            best_index = min(
+                improvement_potentials, key=improvement_potentials.get
+            )
+            num += delta_num[best_index]
+            weight_within_side += delta_denom[best_index]
+            for j in G.neighbors(best_index):
+                if check_if_cut(best_index, j, new_cut):
+                    delta_num[j] += 2 * G[best_index][j]["weight"]
+                else:
+                    delta_num[j] -= 2 * G[best_index][j]["weight"]
+                set_improvement_potential(j)
+            delta_num[best_index] = -delta_num[best_index]
+            delta_denom[best_index] = -delta_denom[best_index]
+            set_improvement_potential(best_index)
+            if best_index in new_cut:
+                new_cut.remove(best_index)
+            else:
+                new_cut.append(best_index)
+        else:
+            all_pos = True
+        iters += 1
+
+    return new_cut
+
+
+def similarity(
+    u: int,
+    v: int,
+    cm: pd.DataFrame,
+    missing_char: str,
+    w: Optional[Dict[int, Dict[str, float]]] = None,
+) -> Union[int, float]:
+    """A function to return the number of (non-missing) character/state
+    mutations shared by two samples.
+
+    Args:
+        u: The row index of the character matrix representing the first sample
+        v: The row index of the character matrix representing the second sample
+        cm: The character matrix of observed character states for all samples
+        missing_char: The character representing missing values
+        w: A set of optional weights for edges in the connectivity graph
+    Returns:
+        The number of shared mutations between two samples, weighted or unweighted
+    """
+
+    # TODO Optimize this using masks
+    k = cm.shape[1]
+    if w is None:
+        return sum(
+            [
+                1
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
+    else:
+        return sum(
+            [
+                w[i][cm.iloc[u, i]]
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
```

Comment:
I think we can extend this in two ways:

Firstly, this should be the modified hamming similarity: if two characters are the same, give a +2; if two characters are different and one is zero, give a +1; if two characters are different, give +0. 

Secondly, we can improve this even more using state priors. We can add log-probs of these states so that if two characters are the same we would give + (log(a) + log(b)), etc.